src/ServiceControl/Recoverability/Retrying/Infrastructure/ReturnToSenderDequeuer.cs

```diff
@@ -137,16 +143,17 @@ public Task CreateQueue()
             }
         }
 
-        public void Stop()
+        public Task Stop()
         {
             timer.Dispose();
             endedPrematurelly = true;
-            Task.Run(() => syncEvent.TrySetResult(true));
+            return StopInternal();
         }
 
-        void StopInternal()
+        private async Task StopInternal()
         {
-            Task.Run(() => syncEvent.TrySetResult(true));
+            await Task.Run(() => syncEvent?.TrySetResult(true)).ConfigureAwait(false);
+            await (stopCompletionSource?.Task ?? (Task)Task.FromResult(0)).ConfigureAwait(false);
```

Comment:
I think there is a potential race condition in case `Stop` is called before `Run` is executed (or when executed concurrently), but that's already the case right now so probably not the most critical thing to address?